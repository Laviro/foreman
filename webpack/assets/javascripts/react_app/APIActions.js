import { API_OPERATIONS } from './APIConstants';

export const stopPollingAction = key => ({
  type: API_OPERATIONS.STOP_POLLING,
  payload: { key },
});
