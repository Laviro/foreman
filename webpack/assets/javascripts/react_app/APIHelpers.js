import { API_OPERATIONS, DEFAULT_POLLING } from './APIConstants';

export const pollingActionGenerator = (outputType, APIRequest, polling) => ({
  type: API_OPERATIONS.START_POLLING,
  outputType,
  payload: {
    APIRequest,
    polling,
  },
});

export const startAPIInterval = (APIRequest, ms) => {
  const pollingMiliSec = typeof ms === 'number' ? ms : DEFAULT_POLLING;
  return setInterval(APIRequest, pollingMiliSec);
};

export const stopAPIInterval = id => {
  clearInterval(id);
};
