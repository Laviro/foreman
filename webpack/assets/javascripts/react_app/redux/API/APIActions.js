import { API_OPERATIONS } from './APIConstants';

export const stopPolling = key => ({
  type: API_OPERATIONS.STOP_POLLING,
  key,
});

export const startPolling = (key, APIRequest, polling) => ({
  type: API_OPERATIONS.START_POLLING,
  key,
  payload: {
    APIRequest,
    polling,
  },
});

export default {
  startPolling,
  stopPolling,
};
