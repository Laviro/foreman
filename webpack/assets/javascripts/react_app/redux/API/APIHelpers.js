import { DEFAULT_POLLING } from './APIConstants';

export const startAPIInterval = (APIRequest, ms) => {
  const pollingMiliSec = typeof ms === 'number' ? ms : DEFAULT_POLLING;
  return setInterval(APIRequest, pollingMiliSec);
};

export const stopAPIInterval = id => {
  clearInterval(id);
};

export const registeredPollingException = key =>
  new Error(
    `There is already a polling process running and registered for: ${key}. Clear it or use another key if you wish to create another polling process.`
  );

export const unregisteredPollingException = key =>
  new Error(`Can't find a registered polling process for: ${key}`);
