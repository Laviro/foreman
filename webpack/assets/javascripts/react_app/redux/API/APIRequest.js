import { API } from './';
import { actionTypeGenerator } from './APIActionTypeGenerator';
import { noop } from '../../common/helpers';

export const get = async (
  {
    key,
    url,
    headers = {},
    params = {},
    actionTypes = {},
    handleError = noop,
    payload = {},
  },
  { dispatch }
) => {
  const { REQUEST, SUCCESS, FAILURE } = actionTypeGenerator(key, actionTypes);
  const modifiedPayload = { ...payload, url };
  dispatch({
    type: REQUEST,
    key,
    payload: modifiedPayload,
  });
  try {
    const { data } = await API.get(url, headers, params);
    dispatch({
      type: SUCCESS,
      key,
      payload: modifiedPayload,
      response: data,
    });
  } catch (error) {
    handleError(error);
    dispatch({
      type: FAILURE,
      key,
      payload: modifiedPayload,
      response: error,
    });
  }
};
