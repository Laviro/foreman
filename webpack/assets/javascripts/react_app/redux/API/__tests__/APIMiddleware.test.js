import { APIMiddleware, API_OPERATIONS } from '../';
import { get } from '../APIRequest';
import { startPolling } from '../APIActions';
import { registeredPollingException } from '../APIHelpers';

jest.mock('../APIRequest');
jest.mock('../APIActions');

describe('APIMiddleware', () => {
  it('should not call get request', async () => {
    get.mockImplementation(jest.fn());
    APIMiddleware()(jest.fn())({ type: 'TEST' });
    expect(get.mock.calls).toHaveLength(0);
  });

  it('should call get request', async () => {
    get.mockImplementation(jest.fn());
    APIMiddleware()(jest.fn())({ type: API_OPERATIONS.GET });
    expect(get).toBeCalled();
  });

  it('should pass the API action to next', () => {
    const fakeStore = {};
    const fakeNext = jest.fn();

    const action = { type: API_OPERATIONS.GET };
    APIMiddleware(fakeStore)(fakeNext)(action);
    expect(fakeNext).toHaveBeenCalledWith(action);
  });

  it('should pass the polling action to next', async () => {
    const key = 'SOME_KEY';
    const polling = 3000;
    const fakeNext = jest.fn();
    const fakeState = {
      API_operations: {
        polling: {},
      },
    };
    const fakeStore = {
      ...fakeState,
      getState: () => fakeState,
    };
    const action = { type: API_OPERATIONS.GET, key, polling };
    startPolling.mockImplementation(jest.fn());
    APIMiddleware(fakeStore)(fakeNext)(action);
    expect(startPolling).toBeCalled();
    expect(fakeNext).toMatchSnapshot();
  });

  it('should throw error when trying to create polling with existing key', async () => {
    const key = 'SOME_KEY';
    const polling = 3000;
    const fakeNext = jest.fn();
    const fakeState = {
      API_operations: {
        polling: {
          [key]: 1,
        },
      },
    };
    const fakeStore = {
      ...fakeState,
      getState: () => fakeState,
    };
    const action = { type: API_OPERATIONS.GET, key, polling };
    try {
      APIMiddleware(fakeStore)(fakeNext)(action);
    } catch (error) {
      expect(error.message).toBe(registeredPollingException(key).message);
    }
  });
});
