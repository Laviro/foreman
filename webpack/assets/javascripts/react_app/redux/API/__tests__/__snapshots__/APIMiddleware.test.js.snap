// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`APIMiddleware should pass the polling action to next 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      undefined,
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": undefined,
    },
  ],
}
`;
